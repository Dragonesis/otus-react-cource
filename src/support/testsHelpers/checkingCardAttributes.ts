import { screen } from '@testing-library/react'
import { Unit } from '@/services/models'

export const checkingCardAttributes = (unit: Unit) => {
  const card = screen.queryAllByLabelText('Карта ' + unit.name)[0]
  expect(card).toBeInTheDocument()
  expect(screen.getByText(unit.name)).toBeInTheDocument()
  expect(screen.getByText('Сила: ' + unit.power)).toBeInTheDocument()
  expect(screen.getByText('Защита: ' + unit.defense)).toBeInTheDocument()
  expect(screen.getByText('Класс: ' + unit.className)).toBeInTheDocument()
}
