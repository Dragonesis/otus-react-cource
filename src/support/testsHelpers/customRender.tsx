import React from 'react'
import { render } from '@testing-library/react'
import { StoreContext } from '@/services/adapters/store'
import { DeckOfCard, Unit } from '@/services/models'

export const customRender = (
  children: React.ReactNode,
  {
    providerProps,
    ...renderOptions
  }: {
    providerProps: {
      cardDeck: DeckOfCard | undefined
      cardsInHand: Unit[]
      setDeckOfCard: (arg: DeckOfCard) => DeckOfCard
      setCardsInHand: (arg: Unit[]) => Unit[]
    }
  },
) => {
  return render(<StoreContext.Provider value={providerProps}>{children}</StoreContext.Provider>, renderOptions)
}
