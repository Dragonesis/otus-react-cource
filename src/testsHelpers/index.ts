export { customRender } from './customRender'
export { checkingCardAttributes } from './checkingCardAttributes'
