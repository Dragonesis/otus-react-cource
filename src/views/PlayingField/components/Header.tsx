import React, { FC } from 'react'
import styled from '@emotion/styled'
import { Container, Button } from '@/ui'
import { useStore } from '@/services/adapters/store'
import { deckOfCardList } from '@/services/mocks'
import { getDeckOfCard } from '@/application'

export interface HeaderProps {
  name: string
}

export const Header: FC<HeaderProps> = ({ name }) => {
  const { setDeckOfCard, setCardsInHand } = useStore()
  
  return (
    <Core>
      <ModContainer>
        <Name>{name}</Name>
        <ActionNewGame 
          variant='light'
          fixSize='s'
          onClick={() => {
            setDeckOfCard(getDeckOfCard(deckOfCardList))
            setCardsInHand([])
          }}
        >Новая игра</ActionNewGame>
      </ModContainer>
    </Core>
  )
}

const Core = styled.div`
  background-color: #3a3a3a;
  padding-top: 10px;
  padding-bottom: 10px;
`

const ModContainer = styled(Container)`
  display: flex;
  align-items: center;
  justify-content: space-between;
`

const Name = styled.div`
  font-size: 18px;
  color: #fff;
`

const ActionNewGame = styled(Button)`
  font-size: 18px;
`
