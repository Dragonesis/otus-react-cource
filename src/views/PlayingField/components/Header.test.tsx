import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import { customRender } from '@/support/testsHelpers'

import { Header } from './Header'

describe('### Header', () => {
  const setDeckOfCard = jest.fn()
  const setCardsInHand = jest.fn()

  const providerProps = {
    cardDeck: undefined,
    cardsInHand: [],
    setDeckOfCard,
    setCardsInHand,
  }

  test('renders correctly', () => {
    const { asFragment } = render(<Header name={'Test'} />)
    expect(asFragment()).toMatchSnapshot()
  })

  test('start new game', () => {  
    customRender(<Header name={'Test'} />, { providerProps })
    const action = screen.getByText('Новая игра')
    fireEvent.click(action)
    expect(setDeckOfCard).toHaveBeenCalled()
    expect(setCardsInHand).toHaveBeenCalled()
  })

})