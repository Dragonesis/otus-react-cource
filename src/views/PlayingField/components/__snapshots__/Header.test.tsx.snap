// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`### Header renders correctly 1`] = `
<DocumentFragment>
  .emotion-0 {
  background-color: #3a3a3a;
  padding-top: 10px;
  padding-bottom: 10px;
}

.emotion-1 {
  max-width: 1000px;
  width: 100%;
  margin-left: auto;
  margin-right: auto;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
}

.emotion-2 {
  font-size: 18px;
  color: #fff;
}

.emotion-3 {
  border-radius: 6px;
  width: 100%;
  max-width: 180px;
  height: 36px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 16px;
  background-color: #fff;
  color: #3a3a3a;
  font-size: 18px;
}

.emotion-3:hover,
.emotion-3:focus {
  border-color: #303030;
}

.emotion-3.size-s {
  height: 36px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 16px;
}

.emotion-3.size-m {
  height: 48px;
  padding-left: 10px;
  padding-right: 10px;
  font-size: 18px;
}

.emotion-3.size-l {
  height: 60px;
  padding-left: 16px;
  padding-right: 16px;
  font-size: 20px;
}

<div
    class="emotion-0"
  >
    <div
      class="emotion-1"
    >
      <div
        class="emotion-2"
      >
        Test
      </div>
      <button
        class="emotion-3"
      >
        Новая игра
      </button>
    </div>
  </div>
</DocumentFragment>
`;
